lib/charms/kubernetes_charm_libraries/v0/volumes.py

```diff
@@ -0,0 +1,415 @@
+
+        Args:
+            statefulset_name: Statefulset name.
+            requested_volumes: Iterable of requested volumes
+
+        Returns:
+            bool: Whether the statefulset has the expected requested volumes.
+        """
+        try:
+            statefulset = self.client.get(
+                res=StatefulSet, name=statefulset_name, namespace=self.namespace
+            )
+        except ApiError as e:
+            if e.status.reason == "Unauthorized":
+                logger.debug("kube-apiserver not ready yet")
+            else:
+                raise KubernetesRequestedVolumesError(
+                    f"Could not get statefulset {statefulset_name}"
+                )
+            return False
+        return self._statefulset_volumes_are_patched(
+            statefulset_spec=statefulset.spec,  # type: ignore[attr-defined]
+            requested_volumes=requested_volumes,
+        )
+
+    @staticmethod
+    def _statefulset_volumes_are_patched(
+        statefulset_spec: StatefulSetSpec,
+        requested_volumes: Iterable[RequestedVolume],
+    ) -> bool:
+        """Returns whether a StatefulSet is patched with requested volumes.
+
+        Args:
+            statefulset_spec: StatefulSet spec
+            requested_volumes: Iterable of requested volumes
+
+        Returns:
+            bool
+        """
+        if not statefulset_spec.template.spec.volumes:
+            return False
+        return all(
+            [
+                requested_volume.volume in statefulset_spec.template.spec.volumes
+                for requested_volume in requested_volumes
+            ]
+        )
+
+    def _pod_contains_requested_volumes(
+        self,
+        container_name: str,
+        requested_volumes: Iterable[RequestedVolume],
+        pod: Union[PodTemplateSpec, Pod],
+    ) -> bool:
+        """Returns whether a pod is patched with requested volumes.
+
+        Args:
+            container_name: Container name
+            requested_volumes: Iterable of requested volumes
+            pod: Kubernetes pod object.
+
+        Returns:
+            bool
+        """
+        return self._pod_volumemounts_contain_requested_volumes(
+            containers=pod.spec.containers,
+            container_name=container_name,
+            requested_volumes=requested_volumes,
+        )
+
+    @staticmethod
+    def _pod_volumemounts_contain_requested_volumes(
+        containers: Iterable[Container],
+        container_name: str,
+        requested_volumes: Iterable[RequestedVolume],
+    ) -> bool:
+        """Returns whether container spec contains the expected requested volumes mounts.
+
+        Args:
+            containers: Iterable of Containers
+            container_name: Container name
+            requested_volumes: Requested volumes we expect to be set
+
+        Returns:
+            bool
+        """
+        container = next(container for container in containers if container.name == container_name)
```

Comment:
If the required container could not be found, it will raise StopIteration error.